File: tools/pdf/buchcover-png.mjs
/**
 * Buchcover „Werde Meister deiner Gedanken" als PNG.
 *
 *   node tools/pdf/buchcover-png.mjs        (npm run buchcover)
 *
 * Gibt die FERTIGE Titelseite (Cover) des Buches als PNG aus – exakt dieselbe
 * Gestaltung wie im Buch-PDF, kein Nachbau: Das Skript baut das Buch-HTML mit
 * build-buch.py, löst daraus die `<section class="cover">` (creme Grund,
 * goldenes Gehirn, „Das Buch", Titel/Untertitel, Autor) heraus und rendert
 * genau diese A4-Seite via Chromium in ein PNG.
 *
 * Quelle des Covers: tools/pdf/build-buch.py  (Inhalt aus
 * docs/ebook/werde-meister-deiner-gedanken.md).
 *
 * Ausgabe:  public/buchcover-werde-meister.png   (2481 × 3509 px, A4 ~300 dpi)
 */
import { readFileSync, writeFileSync, existsSync, readdirSync, rmSync } from "node:fs";
import { execFileSync } from "node:child_process";
import { createRequire } from "node:module";
import { fileURLToPath, pathToFileURL } from "node:url";
import { dirname, join } from "node:path";

const HERE = dirname(fileURLToPath(import.meta.url));
const ROOT = join(HERE, "..", "..");
const BUILD = join(HERE, ".build");
const BOOK_HTML = join(BUILD, "book-wmdg.html");
const OUT = join(ROOT, "public/buchcover-werde-meister.png");

// A4 in CSS-Pixeln (96 dpi) – das Cover ist in build-buch.py 210mm × 297mm.
const CSS_W = Math.ceil((210 / 25.4) * 96); // 794
const CSS_H = Math.ceil((297 / 25.4) * 96); // 1123
const SCALE = 3.125; // → 2481 × 3509 px (A4 bei ~300 dpi)

// 1) Buch-HTML sicherstellen (mit dem vorhandenen Generator bauen).
if (!existsSync(BOOK_HTML)) {
  console.log("• baue Buch-HTML (build-buch.py) …");
  execFileSync("python3", [join(HERE, "build-buch.py")], { stdio: "inherit" });
}

// 2) Nur die Cover-Section + den <head> (Fonts/Styles) herauslösen.
const book = readFileSync(BOOK_HTML, "utf8");
const head = book.slice(book.indexOf("<head>") + 6, book.indexOf("</head>"));
const i = book.indexOf('<section class="cover">');
const j = book.indexOf("</section>", i) + "</section>".length;
if (i < 0 || j < 0) throw new Error("Cover-Section nicht in book-wmdg.html gefunden.");
const cover = book.slice(i, j);

// Standalone-Dokument: exakt eine A4-Coverseite, creme Grund (deckt Rundungen).
const doc = `<!doctype html><html><head>${head}
<style>html,body{margin:0;padding:0;background:#f6f4ee}
.cover{break-after:auto!important}</style></head><body>${cover}</body></html>`;

const tmp = join(BUILD, ".tmp-buchcover.html");
writeFileSync(tmp, doc);

// 3) Rendern: bevorzugt Playwright (elementgenau), sonst Chromium-Headless.
const require = createRequire(import.meta.url);
function findChrome() {
  if (process.env.CHROME_BIN && existsSync(process.env.CHROME_BIN)) return process.env.CHROME_BIN;
  try {
    const p = require("playwright").chromium.executablePath();
    if (p && existsSync(p)) return p;
  } catch {}
  for (const r of [process.env.PLAYWRIGHT_BROWSERS_PATH, "/opt/pw-browsers"].filter(Boolean)) {
    try {
      for (const d of readdirSync(r)) {
        if (d.startsWith("chromium")) {
          const p = join(r, d, "chrome-linux/chrome");
          if (existsSync(p)) return p;
        }
      }
    } catch {}
  }
  throw new Error("Kein Chromium/Chrome gefunden. Führe aus:  npx playwright install chromium");
}

async function renderWithPlaywright() {
  const { chromium } = require("playwright");
  const browser = await chromium.launch({ executablePath: findChrome() });
  const page = await browser.newPage({
    viewport: { width: CSS_W, height: CSS_H },
    deviceScaleFactor: SCALE,
  });
  await page.goto(pathToFileURL(tmp).href, { waitUntil: "networkidle" });
  const el = await page.$(".cover");
  await el.screenshot({ path: OUT });
  await browser.close();
}

function renderWithChromeHeadless() {
  execFileSync(
    findChrome(),
    [
      "--headless=new",
      "--no-sandbox",
      "--hide-scrollbars",
      `--force-device-scale-factor=${SCALE}`,
      `--window-size=${CSS_W},${CSS_H}`,
      "--default-background-color=f6f4eeff",
      "--virtual-time-budget=8000",
      `--screenshot=${OUT}`,
      pathToFileURL(tmp).href,
    ],
    { stdio: "ignore" }
  );
}

try {
  await renderWithPlaywright();
} catch {
  renderWithChromeHeadless();
}
try { rmSync(tmp, { force: true }); } catch {}
console.log("✓ public/buchcover-werde-meister.png (" + Math.round(CSS_W * SCALE) + "×" + Math.round(CSS_H * SCALE) + ")");

File: tools/pdf/buchcover-png.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# Buchcover „Werde Meister deiner Gedanken" als PNG.
#
#   python3 tools/pdf/buchcover-png.py        (npm run buchcover)
#
# Gibt die FERTIGE Titelseite (Cover) des Buches als PNG aus – exakt dieselbe
# Gestaltung wie im Buch-PDF, kein Nachbau.
#
# Warum über PDF? Die Cover-Seite ist auf A4 (210×297mm) im DRUCK-Layout gebaut
# (`@page`/`.cover{height:297mm}`, Fuß per `margin-top:auto` an die Seitenkante
# gesetzt). Am Bildschirm wächst die Seite mit dem Inhalt und der Autoren-Fuß
# würde abgeschnitten. Deshalb rendern wir – wie das Buch selbst – im DRUCK-Modus
# zu einer 1-seitigen A4-PDF und rastern diese zu PNG.
#
# Ablauf:
#   1. Buch-HTML mit build-buch.py sicherstellen (`.build/book-wmdg.html`).
#   2. Daraus <head> (Fonts/Styles) + die <section class="cover"> herauslösen.
#   3. Chromium `--print-to-pdf` → genau eine A4-Seite (das Cover).
#   4. PyMuPDF rastert Seite 1 bei DPI (Standard 300) → PNG.
#
# Ausgabe:  content/pdf/Werde-Meister-deiner-Gedanken-Cover.png (2479×3508 @ 300 dpi)
#           Bewusst NICHT unter public/ (Next.js würde es sonst direkt ausliefern) –
#           es liegt neben dem Buch-PDF im geschützten content/pdf/.
#
# Voraussetzung: PyMuPDF  ->  pip install pymupdf
import os, sys, glob, subprocess, tempfile

HERE = os.path.dirname(os.path.abspath(__file__))
ROOT = os.path.abspath(os.path.join(HERE, "..", ".."))
BUILD = os.path.join(HERE, ".build")
BOOK_HTML = os.path.join(BUILD, "book-wmdg.html")
OUT = os.path.join(ROOT, "content", "pdf", "Werde-Meister-deiner-Gedanken-Cover.png")
DPI = int(os.environ.get("BUCHCOVER_DPI", "300"))


def find_chrome():
    c = os.environ.get("CHROME_BIN")
    if c and os.path.exists(c):
        return c
    for base in [os.environ.get("PLAYWRIGHT_BROWSERS_PATH"), "/opt/pw-browsers"]:
        if not base:
            continue
        for d in sorted(glob.glob(os.path.join(base, "chromium*"))):
            p = os.path.join(d, "chrome-linux", "chrome")
            if os.path.exists(p):
                return p
    from shutil import which
    for name in ("google-chrome", "chromium", "chromium-browser", "chrome"):
        p = which(name)
        if p:
            return p
    raise SystemExit("Kein Chromium/Chrome gefunden. Setze CHROME_BIN oder installiere Chromium.")


def main():
    # 1) Buch-HTML sicherstellen.
    if not os.path.exists(BOOK_HTML):
        print("• baue Buch-HTML (build-buch.py) …")
        subprocess.run([sys.executable, os.path.join(HERE, "build-buch.py")], check=True)

    # 2) <head> + Cover-Section herauslösen.
    html = open(BOOK_HTML, encoding="utf-8").read()
    head = html[html.find("<head>") + 6 : html.find("</head>")]
    i = html.find('<section class="cover">')
    j = html.find("</section>", i) + len("</section>")
    if i < 0 or j < 0:
        raise SystemExit("Cover-Section nicht in book-wmdg.html gefunden.")
    cover = html[i:j]

    doc = (
        "<!doctype html><html><head>" + head +
        "<style>html,body{margin:0;padding:0}</style></head><body>" + cover + "</body></html>"
    )

    os.makedirs(BUILD, exist_ok=True)
    os.makedirs(os.path.dirname(OUT), exist_ok=True)
    tmp_html = os.path.join(BUILD, ".tmp-buchcover.html")
    tmp_pdf = os.path.join(BUILD, ".tmp-buchcover.pdf")
    open(tmp_html, "w", encoding="utf-8").write(doc)

    # 3) Chromium: Druck-Layout → 1-seitige A4-PDF (wie das Buch).
    chrome = find_chrome()
    subprocess.run(
        [
            chrome, "--headless=new", "--no-sandbox", "--no-pdf-header-footer",
            "--virtual-time-budget=15000",
            "--print-to-pdf=" + tmp_pdf,
            "file://" + tmp_html,
        ],
        check=True, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL,
    )

    # 4) PDF → PNG rastern.
    try:
        import pymupdf
    except Exception:
        try:
            import fitz as pymupdf  # ältere PyMuPDF-Versionen
        except Exception:
            raise SystemExit("PyMuPDF fehlt. Installiere es mit:  pip install pymupdf")

    pdf = pymupdf.open(tmp_pdf)
    zoom = DPI / 72.0
    pix = pdf[0].get_pixmap(matrix=pymupdf.Matrix(zoom, zoom))
    pix.save(OUT)
    pdf.close()

    for f in (tmp_html, tmp_pdf):
        try:
            os.remove(f)
        except OSError:
            pass

    rel = os.path.relpath(OUT, ROOT)
    print("✓ %s (%d×%d, A4 @ %d dpi)" % (rel, pix.width, pix.height, DPI))


if __name__ == "__main__":
    main()
